### PR TITLE
Introduce vk() and sc() key chord specifiers

### LINKED ISSUE #7539: Default bound key combinations are not working as expected on international keyboard layouts
# Environment

```none
Windows build number: Microsoft Windows [Version 10.0.19041.450]
Windows Terminal version (if applicable): 1.2.2381.0

```
+ *Swedish* keyboard layout

# Steps to reproduce

invoke the Font Size Increase command <kbd>ctrl</kbd>+<kbd>=</kbd> on a *Swedish* keyboard layout.

(hint: that would be <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>0</kbd>)

# Expected behaviour

Font size increases one step.

# Actual behaviour

Nothing happens.

# Reflections

- The new Command Palette (that I am already a huge fan of, despite it's flaws) insists that Increase Font Size is bound to <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>0</kbd> (bonus bug?)

- Referring to the **Modifier keys** section in http://docs.microsoft.com/en-us/windows/terminal/customize-settings/key-bindings#key-binding-properties it appears that all the characters that are <kbd>shift</kbd>:ed on an American keyboard layout are invalid bindings. This theory is also widely supported by the fact that Windows Terminal binds <kbd>ctrl</kbd>+<kbd>shift</kbd>+*numeral* to *newTab* and <kbd>ctrl</kbd>+<kbd>alt</kbd>+*numeral* to *switchToTab*.

- International keyboards often move the special characters around on the keyboard. (In the case with the Swedish keyboard layout, to fit in the three extra vowels in our 29 letter alphabet.)

- It appears that the problem is how Terminal deals with characters that are bound to the numeric keys.

### LINKED ISSUE #10203: Make Quake mode work with a German keyboard layout
Should the quake mode work with a German keyboard layout out of the box? The \` key has a special handling on the German keyboard, because it is used to handle accents and is not processed immediately. So \` + a produces à. If you want to type a \`, you have to press the \` + [space].

_Originally posted by @lafe in https://github.com/microsoft/terminal/issues/8888#issuecomment-848573279_

### PR DESCRIPTION
**Pull request by lhecker**
This commit introduces an alternative to specifying key bindings as a combination of key modifiers and a character. It allows you to specify an explicit virtual key as `vk(nnn)`.
Additionally this commit makes it possible to bind actions to scan codes. As scan code 41 appears to be the button below the Escape key on virtually all keyboards, we'll be able to bind the quake mode hotkey to `win+sc(41)` and have it work consistently across most if not all keyboard layouts.

## PR Checklist
* [x] Closes #7539, Closes #10203
* [x] I work here
* [x] Tests added/passed

## Validation Steps Performed

The following was tested both on US and DE keyboard layouts:
* Ctrl+, opens settings ✔️
* Win+` opens quake mode window ✔️
* Ctrl+plus/minus increase/decrease font size ✔️

### CODE REVIEW

File: src/cascadia/TerminalControl/KeyChord.idl
Comment (lhecker) [inline src/cascadia/TerminalControl/KeyChord.idl:11]: I didn't want to add a 5th constructor, so I removed 1 instead.

File: src/cascadia/TerminalSettingsModel/ActionMap.cpp
Comment (lhecker) [inline src/cascadia/TerminalSettingsModel/ActionMap.cpp:699]: We'll now have to run `_GetActionByKeyChordInternal` twice, since the given `keys` might have both a vkey as well as a scan code set, but the internal hashmaps have only key bindings loaded that contain either of the two.

I used to have a comment explaining this here, but I must've lost it at some point. I'll re-add one.

File: src/cascadia/TerminalSettingsModel/KeyChordSerialization.cpp
Comment (lhecker) [inline src/cascadia/TerminalSettingsModel/KeyChordSerialization.cpp:188]: Note to self: `sc(41)`

File: src/cascadia/TerminalSettingsModel/defaults-universal.json
Comment (lhecker) [inline src/cascadia/TerminalSettingsModel/defaults-universal.json:174]: This fixes zooming for non-US keyboards.

File: src/cascadia/TerminalSettingsModel/defaults.json
Comment (lhecker) [inline src/cascadia/TerminalSettingsModel/defaults.json:301]: This fixes quake mode for non-US keyboards.

File: src/cascadia/WindowsTerminal/AppHost.cpp
Comment (lhecker) [inline src/cascadia/WindowsTerminal/AppHost.cpp:687]: By casting the `cmd` prematurely, we won't have to cast it again in `AppHost::_GlobalHotkeyPressed`.
Additionally we won't have to hold onto the `_hotkeyActions` map.

I explain this change a bit more in `IslandWindow::RegisterHotKey` below.

File: src/cascadia/WindowsTerminal/AppHost.cpp
Comment (lhecker) [inline src/cascadia/WindowsTerminal/AppHost.cpp:714]: As you can see here, a lot of casting is gone without the `_hotkeyActions` map.
Otherwise this code is still the same - just less indented.

File: src/cascadia/WindowsTerminal/AppHost.h
Comment (lhecker) [inline src/cascadia/WindowsTerminal/AppHost.h:33]: `_useNonClientArea` wasn't initialized during construction. I moved them down so this class uses less padding.

File: src/cascadia/WindowsTerminal/IslandWindow.cpp
Comment (lhecker) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:1000]: The idea behind the change to the interface of `IslandWindow::RegisterHotKey` is twofold:
* By passing the hotkey not as a vector but rather one by one, we can get rid of the `std::vector` of key chords in the AppHost
* This allows IslandWindow to manage the hotkey ID itself! With `_nextHotkeyIndex` in place it's impossible to accidentally call `RegisterHotKey` twice with the same ID, whereas this was possible with the previous API. This change basically improves the "encapsulation" of the IslandWindow class and prevents its (unsafe-use-) details from leaking.

File: tools/GenerateHeaderForJson.ps1
Comment (lhecker) [inline tools/GenerateHeaderForJson.ps1:27]: I had to add support for `"(` and `)"` escaping (now it uses `R"#(...)#"` instead), and I thought I could clean this file up a bit. 👍

File: src/cascadia/WindowsTerminal/IslandWindow.cpp
Comment (lhecker) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:996]: By passing the hotkey not as a vector but rather one by one, we can get rid of the `std::vector` of key chords in the AppHost.

File: src/cascadia/WindowsTerminal/IslandWindow.h
Comment (zadjii-msft) [inline src/cascadia/WindowsTerminal/IslandWindow.h:84]: fwiw I don't think `1` is a sensible value here - `96` would be more appropriate, but leaving it as `1` would definitely do something obviously wrong if we ever tried using it before initializing

File: src/cascadia/WindowsTerminal/IslandWindow.h
Comment (zadjii-msft) [inline src/cascadia/WindowsTerminal/IslandWindow.h:85]: is this unused now?

File: src/cascadia/TerminalSettingsModel/ActionMap.cpp
Comment (zadjii-msft) [inline src/cascadia/TerminalSettingsModel/ActionMap.cpp:518]: imo the original is clearer to read - is there a specific reason to do it this way?

File: src/cascadia/TerminalSettingsModel/ActionMap.cpp
Comment (zadjii-msft) [inline src/cascadia/TerminalSettingsModel/ActionMap.cpp:691]: would it be tremendously dumb to oneline this as 

```c++
        return _GetActionByKeyChordInternal({ modifiers, keys.Vkey(), 0 }).value_or( _GetActionByKeyChordInternal({ modifiers, 0, keys.ScanCode() }).value_or(nullptr));
```

**Review by zadjii-msft — COMMENTED** [review]
meh, these are all nits, but the PR is still in draft, so I'll hold off ✔️-ing until the final version

**Comment by KalleOlaviNiemitalo** [comment]
Should this be added to profiles.schema.json too?

File: src/cascadia/TerminalSettingsModel/ActionMap.cpp
Comment (lhecker) [reply src/cascadia/TerminalSettingsModel/ActionMap.cpp:518]: Yeah, the compiler can fail to remove the error checking and exception handling from a call to `.value()` even if you previously already checked whether `std::optional` contains something. The `operator*` will always return a direct zero-overhead reference into the optional's contents instead.

Additionally if we ever end up enabling clang-tidy in our project it'll nag for using `value()` instead of `operator*` in such situations. I think it's fairly likely that we'll use clang-tidy in the future and as such I think it's alright to change such case prematurely (if we're touching the code anyways).

In this case though, it's a function I didn't exactly change, so I can revert it if you want to.

File: src/cascadia/TerminalSettingsModel/ActionMap.cpp
Comment (lhecker) [reply src/cascadia/TerminalSettingsModel/ActionMap.cpp:691]: It would certainly be nice! But it would also eagerly check both variants, before returning either, and I'm not sure if that's a good idea...

**Comment by lhecker** [comment]
@KalleOlaviNiemitalo Yeah I can do that. 👍
The space in the description field is rather limited though, so I've planned to add the actual, exhaustive, description to the online docs only.

**Comment by KalleOlaviNiemitalo** [comment]
Rather than the description, I meant this regular expression <https://github.com/microsoft/terminal/blob/84e30bcd3a63d5f6207bee5af43e43600dc328e6/doc/cascadia/profiles.schema.json#L7>

**Comment by lhecker** [comment]
@KalleOlaviNiemitalo Oh wow that's a rather... comprehensive regex. 😄
I'm not sure if it's worth keeping it _that_ complex, since it would get out of sync every time the `VKEY_NAME_PAIRS` list is changed in the code. (And it uses backtracking...)

For now I'll just remove the backtracking from the regex and update it with the 5 new key chord types.

File: src/cascadia/LocalTests_SettingsModel/DeserializationTests.cpp
Comment (lhecker) [inline src/cascadia/LocalTests_SettingsModel/DeserializationTests.cpp:1976]: I find these tests interesting... They haven't worked since they were introduced (due to the incorrect capitalization). Should I just remove them?

**Review by lhecker — COMMENTED** [review]
Remaining TODOs:
* Add a comment to `ActionMap::GetActionByKeyChord`
* Add a unit test for `vk(nnn)` / `sc(nnn)` parsing

The remaining implementation in this PR is "ready" from my side though.

File: src/cascadia/LocalTests_SettingsModel/DeserializationTests.cpp
Comment (zadjii-msft) [reply src/cascadia/LocalTests_SettingsModel/DeserializationTests.cpp:1976]: nah, this is just a good catch, thanks!

File: src/cascadia/TerminalSettingsModel/ActionMap.cpp
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsModel/ActionMap.cpp:518]: ok, good enough for me!

File: doc/cascadia/profiles.schema.json
Comment (zadjii-msft) [inline doc/cascadia/profiles.schema.json:7]: If there's an actually correct regex for all the things this accepts, I won't be shocked, but I don't think it's valuable to try and optimize this more past this point IMO.

File: src/til/ut_til/string.cpp
Comment (zadjii-msft) [inline src/til/ut_til/string.cpp:61]: We may want to file a backlog follow-up to add this test to onefuzz after #10431 merges

**Review by zadjii-msft — APPROVED** [review]

File: doc/cascadia/profiles.schema.json
Comment (lhecker) [reply doc/cascadia/profiles.schema.json:7]: Yeah I made it less correct than it currently is, because the previous regexp was really really hard to understand. 😅

File: src/cascadia/TerminalSettingsModel/KeyChordSerialization.cpp
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/KeyChordSerialization.cpp:8]: huh which thing required charconv again?

File: src/cascadia/WindowsTerminal/AppHost.h
Comment (DHowett) [inline src/cascadia/WindowsTerminal/AppHost.h:28]: this does somewhat violate the design where any action could have been promoted to be "global" :/

Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/AppHost.h:28]: I don't think we'll need it, at least not for a while. Methinks the event horizon for "slam arbitrary actions into the tray flyout" is decently far away.

Comment (lhecker) [reply src/cascadia/WindowsTerminal/AppHost.h:28]: I just changed it because I found the code hard to read and I applied the YAGNI principle to it. I find it easier to understand now, as less casts/checks are necessary.
But of course I can't know how others about it. If you feel like I should revert it I'll gladly do so!
I'll wait a bit before I merge it. :)

File: src/cascadia/TerminalSettingsModel/KeyChordSerialization.cpp
Comment (lhecker) [reply src/cascadia/TerminalSettingsModel/KeyChordSerialization.cpp:8]: I initially tried to convert it from wchar_t to char, but I was so annoyed by that, that I `stroul` which even had support for `0x` prefixes (nice!). But that one again doesn't work with non-null-terminating strings. This experience and utter lack of support for wchar_t number parsing was so infuriating for me I just built my own function in the end. 😄

File: src/cascadia/WindowsTerminal/AppHost.h
Comment (lhecker) [reply src/cascadia/WindowsTerminal/AppHost.h:28]: Of course we can also just convert the type of this vector to be `std::vector<Command>` in the future and retain the same advantages. 🙂

**Comment by DarthJahus** [comment]
👏 This is a huge win for key bindings in general. I wish all video games switched to scan codes instead of being layout dependent.

**Pull request merged**
